validation: followups for de-duplication of packages

glozow opened the pull request:
This addresses some comments from review on e12fafda2dfbbdf63f125e5af797ecfaa6488f66 from #22674.

- Improve documentation about de-duplication: [comment](https://github.com/bitcoin/bitcoin/pull/22674/files#r770156708)
- Fix code looking up same-txid-different-witness transaction in mempool: [comment](https://github.com/bitcoin/bitcoin/pull/22674/files#r770804029)
- Improve the interface for when a same-txid-different-witness transaction is swapped: [comment](https://github.com/bitcoin/bitcoin/pull/22674/files#r770782822)
- Add a test for witness swapping: [comment](https://github.com/bitcoin/bitcoin/pull/22674/files#r770804029)
- Add a test for packages with a mix of duplicate/different witness/new parents: [comment](https://github.com/bitcoin/bitcoin/pull/22674#discussion_r773037608)
- Fix issue with not notifying `CValidationInterface` when there's a partial submission due to fail-fast: [comment](https://github.com/bitcoin/bitcoin/pull/22674#discussion_r773013162)

glozow committed:
[packages] return DIFFERENT_WITNESS for same-txid-different-witness tx

The previous interface required callers to guess that the tx had been
swapped and look up the tx again by txid to find a `MEMPOOL_ENTRY`
result. This is a confusing interface.

Instead, explicitly tell the caller that this transaction was
`DIFFERENT_WITNESS` in the result linked to the mempool entry's wtxid.
This gives the caller all the information they need in 1 lookup, and
they can query the mempool for the other transaction if needed.

glozow committed:
[doc] more detailed explanation for deduplication

glozow committed:
[unit test] different witness in package submission

If/when witness replacement is implemented in the future, this test case
can be easily replaced with witness replacement tests.

t-bast commented on src/test/txpackage_tests.cpp:
```diff
@@ -327,4 +327,122 @@ BOOST_FIXTURE_TEST_CASE(package_submission_tests, TestChain100Setup)
+        BOOST_CHECK(it_parent2_deduped != submit_witness2.m_tx_results.end());
+        BOOST_CHECK(it_parent2_deduped->second.m_result_type == MempoolAcceptResult::ResultType::MEMPOOL_ENTRY);
+        BOOST_CHECK(it_child2 != submit_witness2.m_tx_results.end());
+        BOOST_CHECK(it_child2->second.m_result_type == MempoolAcceptResult::ResultType::WITNESS_SWAPPED);
```
Can you check that the wtxid inside WITNESS_SWAPPED is the wtxid of child1?

t-bast commented on src/test/txpackage_tests.cpp:
```diff
@@ -327,4 +327,122 @@ BOOST_FIXTURE_TEST_CASE(package_submission_tests, TestChain100Setup)
+        BOOST_CHECK(!m_node.mempool->exists(GenTxid::Wtxid(ptx_child2->GetWitnessHash())));
+    }
+
+    // Try submitting Package1{child2, grandchild} where child2 is same-txid-different-witness as
+    // the in-mempool transaction, child1. Since child1 exists in the mempool and its outputs are
+    // available, child2 should be swapped and granchild should be accepted.
+    //
+    // This tests a potential censorship vector in which an attacker broadcasts a competing package
+    // where a parent's witness is mutated. The honest package should be accepted despite the fact
+    // that we don't allow witness replacement.
```
:clap:

t-bast reviewed (approved):
Tiny nit about the test, but LGTM.
ACK https://github.com/bitcoin/bitcoin/pull/23804/commits/fcc2f91ed48b748c19cb99fad40f57de6cd3a445

achow101 commented:
ACK fcc2f91ed48b748c19cb99fad40f57de6cd3a445

ariard reviewed (commented):
ACK fcc2f91

Thanks for taking the comments from #22674!

jnewbery commented on src/test/txpackage_tests.cpp:
```diff
@@ -327,4 +327,122 @@ BOOST_FIXTURE_TEST_CASE(package_submission_tests, TestChain100Setup)
+{
+    LOCK(cs_main);
+
+    // Transactions with a same-txid-different-witness transaction in the mempool should be swapped.
```
I think the word swapped is slightly confusing here. It suggests to me that child1 will be replaced ("swapped") with child2 in the mempool.

glozow replied:
Changed result enum to `DIFFERENT_WITNESS` and verb from "swapped" to "ignored" in comments

jnewbery commented on src/test/txpackage_tests.cpp:
```diff
@@ -327,4 +327,122 @@ BOOST_FIXTURE_TEST_CASE(package_submission_tests, TestChain100Setup)
+
+    // Try submitting Package1{child2, grandchild} where child2 is same-txid-different-witness as
+    // the in-mempool transaction, child1. Since child1 exists in the mempool and its outputs are
+    // available, child2 should be swapped and granchild should be accepted.
```
```suggestion
    // available, child2 should be swapped and grandchild should be accepted.
```

jnewbery replied:
I also think the word "swapped" here is confusing.

jnewbery commented on src/test/txpackage_tests.cpp:
```diff
@@ -327,4 +327,122 @@ BOOST_FIXTURE_TEST_CASE(package_submission_tests, TestChain100Setup)
+    witness2.stack.push_back(std::vector<unsigned char>(2));
+    witness2.stack.push_back(std::vector<unsigned char>(witnessScript.begin(), witnessScript.end()));
+
+    CKey child_key;
+    child_key.MakeNewKey(true);
+    CScript child_locking_script = GetScriptForDestination(WitnessV0KeyHash(child_key.GetPubKey()));
```
I think this is slightly confusing terminology, but I realise it's the same as the other tests in this file, so maybe should be left as it is for now. The child transaction's output is locked with a public key, which is signed for with the corresponding private key in the grandchild transaction. I therefore think it'd be clearer if this was named `grandchild_key`, since the entity that creates the grandchild transaction owns this private key.

jnewbery commented on src/validation.h:
```diff
@@ -162,8 +162,13 @@ struct MempoolAcceptResult {
         VALID, //!> Fully validated, valid.
         INVALID, //!> Invalid.
         MEMPOOL_ENTRY, //!> Valid, transaction was already in the mempool.
+        WITNESS_SWAPPED, //!> A same-txid-different-witness tx already exists in the mempool, and was not replaced.
```
Again, I think this terminology is a bit misleading, since swapped suggests the witness was replaced in the mempool.

jnewbery reviewed (commented):
I've left a few minor comments here, but there are a couple of more substantial comments in the original PR here: https://github.com/bitcoin/bitcoin/pull/22674#pullrequestreview-837111576.

glozow committed:
AcceptPackage fixups

No behavior changes, just clarifications.

glozow committed:
[validation] better handle errors in SubmitPackage

Behavior change: don't quit right after LimitMempoolSize() when a
package is partially submitted. We should still send
TransactionAddedToMempool notifications for
transactions that were submitted.

Not behavior change: add a new package validation result for mempool logic errors.

glozow commented:
Addressed comments and added the fixups from @jnewbery's review on #22674 - most have no behavior changes so they're grouped in 1 commit. Another commit fixes the issue where we're not notifying MainSignals when there's a partial submission.

instagibbs commented on src/validation.h:
```diff
@@ -161,8 +161,13 @@ struct MempoolAcceptResult {
+    /** Result type. Present in all MempoolAcceptResults. */
     const ResultType m_result_type;
+
+    // The following field is only present when m_result_type = ResultType::INVALID
```
field seems to exist in either case?

glozow replied:
Ah yeah, not nonexistent, but unset. Default is valid, with all fields empty, so not much meaning.

glozow replied:
Removed this comment

t-bast reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/23804/commits/0300cb10094efcca1832acaa189e365b7bedaec3

glozow committed:
[unit test] package parents are a mix

glozow commented on src/test/txpackage_tests.cpp:
```diff
@@ -327,4 +327,122 @@ BOOST_FIXTURE_TEST_CASE(package_submission_tests, TestChain100Setup)
+    CKey grandchild_key;
+    grandchild_key.MakeNewKey(true);
+    CScript grandchild_locking_script = GetScriptForDestination(WitnessV0KeyHash(grandchild_key.GetPubKey()));
+    auto mtx_grandchild = CreateValidMempoolTransaction(/* input_transaction */ ptx_child2, /* vout */ 0,
```
Done, applied to all the test code that's added in this PR.

glozow commented:
Added more tests for funsies, the rest of the diff is just changing comments.

@ariard @achow101 @jnewbery @instagibbs @t-bast if you have time to take another look? Would be nice to add these fixups with #22674 before v23

achow101 commented:
ACK 3cd7f693d3ed1bb7cf9ba3e0c482174df3684972

t-bast reviewed (approved):
LGTM, ACK https://github.com/bitcoin/bitcoin/pull/23804/commits/3cd7f693d3ed1bb7cf9ba3e0c482174df3684972

instagibbs reviewed (approved):
ACK 3cd7f693d3ed1bb7cf9ba3e0c482174df3684972

ariard commented on src/validation.cpp:
```diff
@@ -1088,7 +1096,6 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
     LimitMempoolSize(m_pool, m_active_chainstate.CoinsTip(),
                      gArgs.GetIntArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000,
                      std::chrono::hours{gArgs.GetIntArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY)});
-    if (!all_submitted) return false;
```
Well technically if this code change now trigger a `TransactionAddedToMempool()`, I think we're changing the views of our `CMainSignals` consumers and that would constitute a behavior change. That said, if I'm correct, I don't think it's worthy to invalidate ACKs just to update commit description.

ariard reviewed (commented):
ACK 3cd7f69

Substantial changes since my last ACK are new commits adding clarifications, improving the errors handling and extending the unit test.

ClaraBara22 commented:
Can I comment?
